lib/graphql/schema/schema_comparator.rb

```diff
@@ -0,0 +1,713 @@
+
+        def find_changes_in_object_type_interfaces(old_type, new_type)
+          old_interfaces = old_type.interfaces.map(&:name)
+          new_interfaces = new_type.interfaces.map(&:name)
+
+          removed = (old_interfaces - new_interfaces).map{ |interface| object_type_interface_removed(new_type, interface) }
+
+          added = (new_interfaces - old_interfaces).map{ |interface| object_type_interface_added(new_type, interface) }
+
+          removed + added
+        end
+
+        def find_changes_in_object_type_fields(old_type, new_type)
+          old_fields = old_type.fields.values.map(&:name)
+          new_fields = new_type.fields.values.map(&:name)
+
+          removed = (old_fields - new_fields).map{ |field| field_removed(new_type, old_type.fields[field]) }
+
+          added = (new_fields - old_fields).map{ |field| field_added(new_type, new_type.fields[field]) }
+
+          changed = (old_fields & new_fields).map { |field|
+            old_field = old_type.fields[field]
+            new_field = new_type.fields[field]
+
+            changes = find_changes_in_fields(old_type, new_type, old_field, new_field)
+
+            # TODO Description changed
+            # TODO Deprecation changed
+            # TODO should differentiate new deprecations from deprecation reason being changed
+            # TODO Find in fields
+
+            changes
+          }.flatten
+
+          removed + added + changed
+        end
+
+        def find_changes_in_fields(old_type, new_type, old_field, new_field)
+          old_field_type = old_field.type.to_s
+          new_field_type = new_field.type.to_s
+
+          type_change = []
+          # changes << field_type_changed(
+
+          argument_changes = find_changes_in_arguments(
+            old_field.arguments,
+            new_field.arguments,
+            removed_method: lambda { |argument| nil }, # TODO
+            added_method: lambda { |argument, breaking_change| object_type_argument_added(new_type, new_field, argument, breaking_change) },
+            description_method: lambda { |argument| object_type_argument_description_changed(new_type, new_field, argument) }, # TODO
+            default_method: lambda { |argument, old_default_value, new_default_value| object_type_argument_default_changed(new_type, new_field, argument, old_default_value, new_default_value) },
+            type_change_method: lambda { |argument, old_argument_type, new_argument_type, breaking_change| object_type_argument_type_changed(new_type, new_field, argument, old_argument_type, new_argument_type, breaking_change) },
+          )
+
+          type_change + argument_changes
+        end
+
+        def find_changes_in_interface_type(old_type, new_type)
+          find_changes_in_object_type_fields(old_type, new_type)
+        end
+
+        def type_removed(type)
+          {
+            type: TYPE_REMOVED,
+            description: "`#{type}` type was removed",
+            breaking_change: true,
+          }
```

Comment:
Is there an advantage to using Hashes here instead of a dedicated object? The nice thing about objects is that the API is a method call, and a typo results in a no-method-error instead of silent failure.